test/test_transforms_v2_refactored.py

```diff
@@ -0,0 +1,597 @@
+    **kwargs,
+):
+    with mock.patch("torch._C._log_api_usage_once", wraps=torch._C._log_api_usage_once) as spy:
+        dispatcher(input, *args, **kwargs)
+
+        spy.assert_any_call(f"{dispatcher.__module__}.{dispatcher.__name__}")
+
+    unknown_input = object()
+    with pytest.raises(TypeError, match=re.escape(str(type(unknown_input)))):
+        dispatcher(unknown_input, *args, **kwargs)
+
+    if check_scripted_smoke:
+        _check_dispatcher_scripted_smoke(dispatcher, input, *args, **kwargs)
+
+    if check_dispatch_simple_tensor:
+        _check_dispatcher_dispatch_simple_tensor(dispatcher, kernel, input, *args, **kwargs)
+
+    if check_dispatch_datapoint:
+        _check_dispatcher_dispatch_datapoint(dispatcher, kernel, input, *args, **kwargs)
+
+    if check_dispatch_pil:
+        _check_dispatcher_dispatch_pil(dispatcher, input, *args, **kwargs)
+
+
+def _check_dispatcher_kernel_signature_match(dispatcher, *, kernel, input_type):
+    dispatcher_signature = inspect.signature(dispatcher)
+    dispatcher_params = list(dispatcher_signature.parameters.values())[1:]
+
+    kernel_signature = inspect.signature(kernel)
+    kernel_params = list(kernel_signature.parameters.values())[1:]
+
+    # We filter out metadata that is implicitly passed to the dispatcher through the input datapoint, but has to be
+    # explicit passed to the kernel.
+    kernel_params = [param for param in kernel_params if param.name not in input_type.__annotations__.keys()]
+
+    dispatcher_params = iter(dispatcher_params)
+    for dispatcher_param, kernel_param in zip(dispatcher_params, kernel_params):
+        try:
+            # In general, the dispatcher parameters are a superset of the kernel parameters. Thus, we filter out
+            # dispatcher parameters that have no kernel equivalent while keeping the order intact.
+            while dispatcher_param.name != kernel_param.name:
+                dispatcher_param = next(dispatcher_params)
+        except StopIteration:
+            raise AssertionError(
+                f"Parameter `{kernel_param.name}` of kernel `{kernel.__name__}` "
+                f"has no corresponding parameter on the dispatcher `{dispatcher.__name__}`."
+            ) from None
+
+        assert dispatcher_param == kernel_param
+
+
+def _check_dispatcher_datapoint_signature_match(dispatcher):
+    dispatcher_signature = inspect.signature(dispatcher)
+    dispatcher_params = list(dispatcher_signature.parameters.values())[1:]
+
+    datapoint_method = getattr(datapoints._datapoint.Datapoint, dispatcher.__name__)
+    datapoint_signature = inspect.signature(datapoint_method)
+    datapoint_params = list(datapoint_signature.parameters.values())[1:]
+
+    # Because we use `from __future__ import annotations` inside the module where `datapoints._datapoint` is
+    # defined, the annotations are stored as strings. This makes them concrete again, so they can be compared to the
+    # natively concrete dispatcher annotations.
```

Comment:
Here is the quick rundown of what is happening:

Currently, if you put annotations on an object, they are stored as actual objects:

```python
def foo(bar: int, baz: float) -> str:
    pass

print(foo.__annotations__)
```

```
{'bar': <class 'int'>, 'baz': <class 'float'>, 'return': <class 'str'>}
```

This is problematic if you need to annotate with something that is not completely defined yet. A very common case (and the one we are hitting here), is a return annotation on a `@classmethod`. For example

```python
class Foo:
    def __new__(cls) -> Foo:
        pass
```

```
NameError: name 'Foo' is not defined
```

The workaround is to stringify the annotation. Since that is quite ugly, there is this line

https://github.com/pytorch/vision/blob/fc83b280387d92fe6aea73f35c1ed4ad1bb739da/torchvision/datapoints/_bounding_box.py#L1

that enables this behavior automagically. For example

```python
from __future__ import annotations

def foo(bar: int, baz: float) -> str:
    pass

print(foo.__annotations__)

class Foo:
    def __new__(cls) -> Foo:
        pass
```

```
{'bar': 'int', 'baz': 'float', 'return': 'str'}
```

This behavior is supposed to become mandatory in a future release of Python, but so far that was [postponed multiple times](https://docs.python.org/3/library/__future__.html#id1).

---

Coming back the test: to be able to compare the annotations, we need to make sure that we either have the proper types on both objects or the corresponding strings.

Reply:
Ah, thanks for the details, that make sense. I'm going to suggest a different comment below with what I understand. In the current comment, I thought "This" was referring to the previous sentence, when in reality it relates to the code block below - that's a big difference. I also think we can leave out the detail about `from __future__ import annotations`.